### PR TITLE
Laboratory work #4, Tasenko Olga- 21FPL1

### PR DESCRIPTION
**Pull request by Tasenko03**

### CODE REVIEW

File: lab_4_summarization_textrank/main.py [hunk 1]
```diff
@@ -160,22 +231,27 @@ def __init__(self) -> None:
...
         Retrieves a number of vertices that are similar (i.e. have similarity score > 0) to the input one
         :param vertex
         :return:
         """
-        pass
+        check_type(vertex, Sentence, False)
+        if vertex not in self._vertices:
+            raise ValueError
+        idx = self._vertices.index(vertex)
+        return len([element for element in self._matrix[idx] if element > 0])
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:254]: [] is redundant

File: lab_4_summarization_textrank/start.py [hunk 2]
```diff
@@ -27,6 +30,19 @@
...
+    similarity_matrix = SimilarityMatrix()
+    similarity_matrix.fill_from_sentences(sentences)
+    text_rank_summarizer = TextRankSummarizer(similarity_matrix)
+    text_rank_summarizer.train()
+    print(text_rank_summarizer.make_summary(10))
+
+    #RESULT = None
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
-    assert RESULT, 'Summaries are not extracted'
+    #assert RESULT, 'Summaries are not extracted'
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/start.py:48]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py (hunk 1 above)
Comment (Tasenko03) [reply lab_4_summarization_textrank/main.py:263]: TypeError raises without [] 

File "D:\2022-2-level-labs\lab_4_summarization_textrank\main.py", line 263, in calculate_inout_score
    return len(element for element in self._matrix[idx] if element > 0)
TypeError: object of type 'generator' has no len()

File: lab_4_summarization_textrank/main.py [hunk 3]
```diff
@@ -84,59 +154,76 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
         """
         Constructs all the necessary attributes
         """
-        pass
+        super().__init__(stop_words, punctuation)
```
Comment (artyomtugaryov) [inline lab_4_summarization_textrank/main.py:157]: I believe it is better to do check first and than call super

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_summarization_textrank/main.py [hunk 4]
```diff
@@ -84,59 +154,76 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
...
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        check_inner_types(tokens, tuple, str, True)
+        new_tokens = (elem for elem in tokens if elem not in self._word2id)
+
+        for ind, element in enumerate(new_tokens, start=1000 + len(self._word2id)):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:213]: move to own attribute and mutate it

File: lab_4_summarization_textrank/main.py [hunk 5]
```diff
@@ -146,7 +233,14 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
     :param other_sequence: a sequence of items
     :return: similarity score
     """
-    pass
+    if not isinstance(sequence, (list, tuple)) or not isinstance(other_sequence, (list, tuple)):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:236]: rework to use own function to check types

File: lab_4_summarization_textrank/main.py [hunk 6]
```diff
@@ -146,7 +233,14 @@ def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[lis
...
+        raise ValueError
+    if len(sequence) == 0 or len(other_sequence) == 0:
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:238]: ?

File: lab_4_summarization_textrank/main.py [hunk 7]
```diff
@@ -184,7 +283,24 @@ def add_edge(self, vertex1: Sentence, vertex2: Sentence) -> None:
...
+            raise ValueError
+
+        for vertex in vertex1, vertex2:
+            if vertex not in self._vertices:
+                self._vertices.append(vertex)
+                self._matrix.append([])
+
+        for edges_list in self._matrix:
+            if len(edges_list) < len(self._vertices):
+                edges_list.extend([0 for _ in range(len(self._vertices) - len(edges_list))])
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:298]: how many 0 are added from time to time

File: lab_4_summarization_textrank/main.py [hunk 8]
```diff
@@ -257,15 +399,20 @@ def get_top_sentences(self, n_sentences: int) -> tuple[Sentence, ...]:
...
 
     def make_summary(self, n_sentences: int) -> str:
         """
         Constructs summary from the most important sentences
         :param n_sentences: number of sentences to include in the summary
         :return: summary
         """
-        pass
+        check_type(n_sentences, int, False)
+        summery = sorted(self.get_top_sentences(n_sentences), key=lambda elem: elem.get_position())
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:414]: summary

File: lab_4_summarization_textrank/main.py [hunk 9]
```diff
@@ -304,7 +477,18 @@ def _find_texts_close_to_keywords(self, keywords: tuple[str, ...], n_texts: int)
         :param n_texts: number of texts to find
         :return: the texts' ids
         """
-        pass
+        check_inner_types(keywords, tuple, str, True)
+        check_type(n_texts, int, False)
+        if isinstance(n_texts, bool):
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:482]: check_type

File: lab_4_summarization_textrank/main.py [hunk 10]
```diff
@@ -304,7 +477,18 @@ def _find_texts_close_to_keywords(self, keywords: tuple[str, ...], n_texts: int)
...
+            raise ValueError
+        top_similar_texts = {}
+        for element in self._knowledge_database.items():
+            number = calculate_similarity(element[1]['keywords'], keywords)
+            top_similar_texts[element[0]] = number
+        if not any(top_similar_texts.values()):
+            raise NoRelevantTextsError
+        sorted_top_similar_texts = dict(sorted(top_similar_texts.items(), key=lambda x: (x[1], x[0]), reverse=True))
```
Comment (demid5111) [inline lab_4_summarization_textrank/main.py:490]: ??

File: lab_4_summarization_textrank/start.py [hunk 11]
```diff
@@ -27,6 +30,27 @@
...
+    text_rank_summarizer = TextRankSummarizer(similarity_matrix)
+    text_rank_summarizer.train()
+    SUMMARY = text_rank_summarizer.make_summary(5)
+    print(SUMMARY)
+
+    # mark 10
+    buddy = Buddy(paths_to_texts, stop_words, tuple(punctuation), idf)
+    for path in paths_to_texts:
+        buddy.add_text_to_database(path)
+    print(buddy.reply('Кто такой Гагарин?'))
```
Comment (demid5111) [inline lab_4_summarization_textrank/start.py:52]: handle errors

**Review by demid5111 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**